### ISSUE TITLE
Template handler does not handle BlockList component templates

### ISSUE THREAD

**Issue by kdx-perbol** [KevinJump/uSync8#203]
Since the introduction of BlockLists in Umbraco, Templates are not only stored in the root Templates folder. The default location for Component templates for BlockLists is Partials/BlockList/Components. But it seems only the root template folder is handled by the template handler.

This means that all component templates are missing from uSync. And this, in turn, means that when importing document types with component templates (typically Element Type templates), uSync will complain about a missing template cshtml file.

Moving settings between environment works anyway, since Umbraco will at runtime find the correct template cshtml file location using the OotB block list renderer (Partials/BlockList/Default.cshtml). But it causes headaches for development since a big **chunk** of templates is not synchronized correctly between developers.

Are there any plans to extend the scope of handled templates?

Or are we missing something?

**Comment by KevinJump** [comment]
Hi, 

we will take a look at that - I think we need to update where it looks.

**Comment by KevinJump** [comment]
Hi, 

Can you give us a bit more info on this setup ? 

If I create a new block list editor data type, and then from within the datatype, "create a new element type" - that creates an element type which doesn't have its own template - (so there is nothing for us to find).

what does your setup look like, i think i might be missing something ?

**Comment by kdx-perbol** [comment]
Hi Kevin, thanks for looking at this.

The templates missing from uSync are templates for "components" aka "partials" aka "BlockList elements" _with_ templates. These templates are used when rendering BlockList elements using the OotB BlockList functionality.

Our file/solution structure is:

```
    Web project root\
      Views\
        Layout.cshtml
        BlogPage.cshtml
        Partials\
          BlockList\
            Components\
              BlogEntry.cshtml
```

And the structure in Umbraco Settings Templating is:

```
    Templates
      Layout
        BlogPage
    Partial Views
      BlockList\
        Components\
          BlogEntry.cshtml
```

This is the standard setup for BlockList components. The folder structure comes OotB with an Umbraco installation, and the Templates structure does as well. BlockLists were introduced in 8.x some time last year iirc.

The process for creating the above is:

* Create a document type with a template for the layout
* Create a document type with a template, under the layout, for the BlogPage
* Create a document type with a template for the BlogEntry
* Move the BlogEntry.cshtml template file to the Partials\BlockList\Components folder in the web project so it can be used by a BlockList
* Modify the BlogPage document type to contain a BlockList with BlogEntry elements
* Modify the BlogPage.cshtml template to render the BlockList

At this point, you have a working page.

In our case, it would be sufficient if uSync handled the Partial Views folder in Umbraco Settings Templating as well as the Templates folder. 

Are we doing something the wrong way? 

Perhaps the Partial Views folder in Templating is just a "bunch of files" and doesn't need to be handled by uSync. But then at least the import process could be aware of BlockList element templates and not show them as missing cshtml files, which they de facto aren't. And since the BlockList elements are document types with templates, we would prefer it if uSync treated them the same way as structure-level document types with templates.

**Comment by KevinJump** [comment]
Hi, 

I can't quite replicate your steps - specifically. 

> Modify the BlogPage document type to contain a BlockList with BlogEntry elements

because you can't add a element to a block list that is not an element type.

![image](https://user-images.githubusercontent.com/431231/106908312-1928ed80-66f7-11eb-9766-d5aa28a962a0.png)


I am guessing you are making a doctype and then making it an element type - this will allow it to be added - but **Document Types of Element type do not have templates**

## I think this is where there is a uSync bug : 

- if you create an document type (without a template) called blog Entry, and ticked the Element type box, then it would all work but it wouldn't have an associated template. 

- however if you create an document type (normally with a template) and then tick the element type box. **a template is created although its not actually used by Umbraco.** - but! because its still in the code somewhere uSync does write it out to the .config file and when you reimport it trys to find the file and it fails. 😢  - we need to fix this , so it doesn't actually attempt to put a template on an element type. 

---
I would say you are better off creating these types from the "Document type without a template" option because you don't actually want the template. 

![image](https://user-images.githubusercontent.com/431231/106909438-36aa8700-66f8-11eb-81e8-dc612a16ca71.png)

When you move the template file from \views to \partial views\blocklist\components you are actually breaking the template link in Umbraco. you can tell this because the template entry will still be there in the list of templates. but when you click on it, it will be blank. 

Writing text in there and saving it, will create a new blogentry.cshtml file in the root of the views folder. 
so you are better not having the template at all 

**To note:** your are seeing this issue in uSync because you have moved the physical template file and within Umbraco it still thinks there should be a file in the root of the views folder. 

so you should probibly create a file directly in the blocklist/components folder. it doesn't need to be a template first. 

As a work around to your existing setup i do the following
(after moving the blogEntry.cshtml) 

* Delete the blogEntry template from the 'Templates' list 
* Go back to the blobEntry docutment type and save it. 

this will have the effect of removing the `<defaultTemplate>` && `<templates>...</templates>` entry in the uSync config file, which is the issue uSync is flagging up.

**Comment by kdx-perbol** [comment]
I understand. But just to be clear, you realize we actually do want a template for the element type, just not in the root Views folder. Since we're using BlockLists to render the element type, the template has to be in the Partials folder, at least when using the OotB BlockList configuration. 

I realize that views for element types aren't really template-templates. 

Your solution will take uSync out of the equation for the template - but we want uSync to be in the equation. :)

**Comment by KevinJump** [comment]
From Umbraco view point the files in partial views/blocklist... are not templates. They are simply files - The partial views folder is a direct representation of the files on disk, there is no DB element to that big of the Umbraco UI.

The block list rendering isn't using templates rather it is 'just' reading the files from a location that follows the convention (as you point out this is written in the `default.cshtml` file in the blocklists folder).

So from that point you don't actually want a `template` in the 'template' folder view - because its not being used by the render process and Umbraco will always think that there is a file for that in the root of views  (if you physically move the file, Umbraco will still look for a file in \views and will in-fact create a new one if you edit the template in the back office (from a UI point of view this is not the file you have moved)

uSync's role in templates to to match the database entries back to the files in the root of the views folder - the actual contents of those files is handles when you copy/deploy the views folder. and as such the partial views folder too - so uSync doesn't need to know about the files inside that bit. 

the "fix" for us is not to warn when the default template is missing on an element, because looking into it, the template reference does still exist in the DB it just no longer shown to the user - so its has minimal effect when its missing and you are in an element type - so we won't mark it as an issue any more. 


when you move the file from \views to \partial views\blocklist.... you are effectivly removing the template - except you are not deleting it in umbraco - and umbraco stores an id for each template in the DB,

**Comment by kdx-perbol** [comment]
I see. Your proposed fix will make our life easier. I guess what would really help is a more template-like solution for BlockLists.

Thank you!

**Comment by KevinJump** [comment]
the remove warning - fix was in 8.8.4 release
